### ISSUE TITLE
Espnet-SE: After using the new loss wrapper and criterion, cannot reproduce the performance in CHiME4/enh1/

### ISSUE THREAD

**Issue by guoyifan97** [espnet/espnet#4298]
Hi there, thanks for your great work. 

I just updated my local code following the PR #3838 (add the loss criterion and the wrapper), and find that the performance degrades on both sdr and asr performance. I also reinstalled the latest version of espnet  and got the same result as the updated code.

**This is the configure file before code update:**
espnet version: 0.10.3a3
torch version: 1.9.1+cu111
##################################################
![image](https://user-images.githubusercontent.com/37100657/164360740-f6b0969f-4580-45e0-9dd1-d1a8e267a0c2.png)
![image](https://user-images.githubusercontent.com/37100657/164360749-5d05e973-5f7c-4493-b827-a6eb2cd08f33.png)

###############################################

**This is the result before update:**
###############################################
et05_simu:
sdr: 15.88
asr performance:  (Use an asr model pretrained on the WSJ, and all asr results are decoded using one model)
cer:  13.9
wer: 27.5
###############################################

**The configure file after the update (also for the newest released espnet)**
espnet version: 202204
torch version: 1.9.1+cu111
###############################################
![image](https://user-images.githubusercontent.com/37100657/164360672-7bc8f262-d46f-4ca9-9631-b3cd46f88b91.png)
![image](https://user-images.githubusercontent.com/37100657/164360687-b7808eb3-6ce7-4f34-9b6b-a1d505d7d2e9.png)
###############################################

**The results (newest espnet):**
###############################################
et05_simu:
sdr: 13.01
asr performance:
cer: 17.7
wer: 34.1
###############################################

**Comment by Emrys365** [comment]
Hi @guoyifan97 , I think the main difference is in the `use_noise_ref` behavior.

In the latest branch, we temporarily disable the loss calculation on the noise data specified in `noise1.scp`, even if `use_noise_ref` is set to True. This is because the new loss wrapper interface makes it a bit tricky to calculate additional losses on non-speech estimates.

**Comment by guoyifan97** [comment]
@Emrys365 Thanks for your reply! But I think the loss types I use before and after the code update are both the MES loss on the spectrum, which don't involve the loss calculations for noisy data.

**Comment by Emrys365** [comment]
Could you check whether `--use_noise_ref` is set to `true` in [egs2/chime4/enh1/run.sh](https://github.com/espnet/espnet/blob/master/egs2/chime4/enh1/run.sh#L32)? If so, an additional loss will be calculated on the estimated noise mask.

**Comment by guoyifan97** [comment]
@Emrys365 Sure! I have checked, and it is set to false.

**Comment by Emrys365** [comment]
Okay, then I will look into this issue and let you know once I find the cause.

**Comment by guoyifan97** [comment]
@Emrys365 Thanks a lot!

**Comment by Emrys365** [comment]
BTW, could you post the `config.yaml` file in your experiment directory?

**Comment by guoyifan97** [comment]
Sure,

##############################
Before update:
`config: conf/tuning/train_enh_beamformer_mvdr_spectrum.yaml
print_config: false
log_level: INFO
dry_run: false
iterator_type: sequence
output_dir: exp/enh_train_enh_beamformer_mvdr_spectrum_raw
ngpu: 1
seed: 0
num_workers: 4
num_att_plot: 3
dist_backend: nccl
dist_init_method: env://
dist_world_size: null
dist_rank: null
local_rank: 0
dist_master_addr: null
dist_master_port: null
dist_launcher: null
multiprocessing_distributed: false
unused_parameters: false
sharded_ddp: false
cudnn_enabled: true
cudnn_benchmark: false
cudnn_deterministic: true
collect_stats: false
write_collected_feats: false
max_epoch: 70
patience: null
val_scheduler_criterion:
- valid
- loss
early_stopping_criterion:
- valid
- loss
- min
best_model_criterion:
-   - valid
    - si_snr
    - max
-   - valid
    - loss
    - min
keep_nbest_models: 1
grad_clip: 5.0
grad_clip_type: 2.0
grad_noise: false
accum_grad: 1
no_forward_run: false
resume: true
init_with_other_model: false
train_dtype: float32
use_amp: false
log_interval: null
use_tensorboard: true
use_wandb: false
wandb_project: null
wandb_id: null
wandb_entity: null
wandb_name: null
wandb_model_log_interval: -1
detect_anomaly: false
pretrain_path: null
init_param: []
ignore_init_mismatch: false
freeze_param: []
num_iters_per_epoch: null
batch_size: 8
valid_batch_size: null
batch_bins: 1000000
valid_batch_bins: null
train_shape_file:
- exp/enh_stats_16k/train/speech_mix_shape
- exp/enh_stats_16k/train/speech_ref1_shape
valid_shape_file:
- exp/enh_stats_16k/valid/speech_mix_shape
- exp/enh_stats_16k/valid/speech_ref1_shape
batch_type: folded
valid_batch_type: null
fold_length:
- 80000
- 80000
sort_in_batch: descending
sort_batch: descending
multiple_iterator: false
chunk_length: 500
chunk_shift_ratio: 0.5
num_cache_chunks: 1024
train_data_path_and_name_and_type:
-   - dump/raw/tr05_simu_isolated_6ch_track/wav.scp
    - speech_mix
    - sound
-   - dump/raw/tr05_simu_isolated_6ch_track/spk1.scp
    - speech_ref1
    - sound
valid_data_path_and_name_and_type:
-   - dump/raw/dt05_simu_isolated_6ch_track/wav.scp
    - speech_mix
    - sound
-   - dump/raw/dt05_simu_isolated_6ch_track/spk1.scp
    - speech_ref1
    - sound
allow_variable_data_keys: false
max_cache_size: 0.0
max_cache_fd: 32
valid_max_cache_size: null
optim: adam
optim_conf:
    lr: 0.001
    eps: 1.0e-08
    weight_decay: 0
scheduler: reducelronplateau
scheduler_conf:
    mode: min
    factor: 0.5
    patience: 1
init: xavier_uniform
model_conf:
    loss_type: spectrum
use_preprocessor: false
encoder: stft
encoder_conf:
    n_fft: 512
    hop_length: 128
    use_builtin_complex: false
separator: wpe_beamformer
separator_conf:
    num_spk: 1
    loss_type: spectrum
    use_wpe: false
    wnet_type: blstmp
    wlayers: 3
    wunits: 300
    wprojs: 320
    wdropout_rate: 0.0
    taps: 5
    delay: 3
    use_dnn_mask_for_wpe: true
    use_beamformer: true
    bnet_type: blstmp
    blayers: 3
    bunits: 512
    bprojs: 512
    badim: 320
    ref_channel: 4
    use_noise_mask: true
    beamformer_type: mvdr_souden
    bdropout_rate: 0.0
decoder: stft
decoder_conf:
    n_fft: 512
    hop_length: 128
required:
- output_dir
version: 0.10.3a3
distributed: false
`
######################

After:
######################
`
config: conf/tuning/train_enh_beamformer_mvdr_new_loss_wrapper_spec_mse.yaml
print_config: false
log_level: INFO
dry_run: false
iterator_type: sequence
output_dir: exp/enh_train_enh_beamformer_mvdr_new_loss_wrapper_spec_mse_raw
ngpu: 1
seed: 0
num_workers: 4
num_att_plot: 3
dist_backend: nccl
dist_init_method: env://
dist_world_size: null
dist_rank: null
local_rank: 0
dist_master_addr: null
dist_master_port: null
dist_launcher: null
multiprocessing_distributed: false
unused_parameters: false
sharded_ddp: false
cudnn_enabled: true
cudnn_benchmark: false
cudnn_deterministic: true
collect_stats: false
write_collected_feats: false
max_epoch: 70
patience: null
val_scheduler_criterion:
- valid
- loss
early_stopping_criterion:
- valid
- loss
- min
best_model_criterion:
-   - valid
    - si_snr
    - max
-   - valid
    - loss
    - min
keep_nbest_models: 1
nbest_averaging_interval: 0
grad_clip: 5.0
grad_clip_type: 2.0
grad_noise: false
accum_grad: 1
no_forward_run: false
resume: true
train_dtype: float32
use_amp: false
log_interval: null
use_matplotlib: true
use_tensorboard: true
use_wandb: false
wandb_project: null
wandb_id: null
wandb_entity: null
wandb_name: null
wandb_model_log_interval: -1
detect_anomaly: false
pretrain_path: null
init_param: []
ignore_init_mismatch: false
freeze_param: []
num_iters_per_epoch: null
batch_size: 8
valid_batch_size: null
batch_bins: 1000000
valid_batch_bins: null
train_shape_file:
- exp/enh_stats_16k/train/speech_mix_shape
- exp/enh_stats_16k/train/speech_ref1_shape
valid_shape_file:
- exp/enh_stats_16k/valid/speech_mix_shape
- exp/enh_stats_16k/valid/speech_ref1_shape
batch_type: folded
valid_batch_type: null
fold_length:
- 80000
- 80000
sort_in_batch: descending
sort_batch: descending
multiple_iterator: false
chunk_length: 500
chunk_shift_ratio: 0.5
num_cache_chunks: 1024
train_data_path_and_name_and_type:
-   - dump/raw/tr05_simu_isolated_6ch_track/wav.scp
    - speech_mix
    - sound
-   - dump/raw/tr05_simu_isolated_6ch_track/spk1.scp
    - speech_ref1
    - sound
valid_data_path_and_name_and_type:
-   - dump/raw/dt05_simu_isolated_6ch_track/wav.scp
    - speech_mix
    - sound
-   - dump/raw/dt05_simu_isolated_6ch_track/spk1.scp
    - speech_ref1
    - sound
allow_variable_data_keys: false
max_cache_size: 0.0
max_cache_fd: 32
valid_max_cache_size: null
optim: adam
optim_conf:
    lr: 0.001
    eps: 1.0e-08
    weight_decay: 0
scheduler: reducelronplateau
scheduler_conf:
    mode: min
    factor: 0.5
    patience: 1
init: xavier_uniform
model_conf:
    stft_consistency: false
    loss_type: mask_mse
    mask_type: null
criterions:
-   name: mse
    conf:
        compute_on_mask: false
        mask_type: PSM^2
    wrapper: fixed_order
    wrapper_conf:
        weight: 1.0
use_preprocessor: false
encoder: stft
encoder_conf:
    n_fft: 512
    hop_length: 128
    use_builtin_complex: false
separator: wpe_beamformer
separator_conf:
    num_spk: 1
    loss_type: spectrum_log
    use_wpe: false
    wnet_type: blstmp
    wlayers: 3
    wunits: 300
    wprojs: 320
    wdropout_rate: 0.0
    taps: 5
    delay: 3
    use_dnn_mask_for_wpe: true
    use_beamformer: true
    bnet_type: blstmp
    blayers: 3
    bunits: 512
    bprojs: 512
    badim: 320
    ref_channel: 3
    use_noise_mask: true
    beamformer_type: mvdr_souden
    bdropout_rate: 0.0
decoder: stft
decoder_conf:
    n_fft: 512
    hop_length: 128
required:
- output_dir
version: '202204'
distributed: false
`

**Comment by Emrys365** [comment]
There are various differences, actually.

  * First, the `loss_type` is changed from `spectrum` to `spectrum_log`.
  * Second, the `ref_channel` is changed from 4 to 3. (This is because it is more common to use CH5 as the reference channel)

```diff
--- conf/tuning/train_enh_beamformer_mvdr_spectrum.yaml
+++ conf/tuning/train_enh_beamformer_mvdr_new_loss_wrapper_spec_mse.yaml
@@ -2,7 +2,7 @@
 log_level: INFO
 dry_run: false
 iterator_type: sequence
-output_dir: exp/enh_train_enh_beamformer_mvdr_spectrum_raw
+output_dir: exp/enh_train_enh_beamformer_mvdr_new_loss_wrapper_spec_mse_raw
 ngpu: 1
 seed: 0
 num_workers: 4
@@ -40,16 +40,17 @@
     - loss
     - min
 keep_nbest_models: 1
+nbest_averaging_interval: 0
 grad_clip: 5.0
 grad_clip_type: 2.0
 grad_noise: false
 accum_grad: 1
 no_forward_run: false
 resume: true
-init_with_other_model: false
 train_dtype: float32
 use_amp: false
 log_interval: null
+use_matplotlib: true
 use_tensorboard: true
 use_wandb: false
 wandb_project: null
@@ -114,7 +115,17 @@
     patience: 1
 init: xavier_uniform
 model_conf:
-    loss_type: spectrum
+    stft_consistency: false
+    loss_type: mask_mse
+    mask_type: null
+criterions:
+-   name: mse
+    conf:
+        compute_on_mask: false
+        mask_type: PSM^2
+    wrapper: fixed_order
+    wrapper_conf:
+        weight: 1.0
 use_preprocessor: false
 encoder: stft
 encoder_conf:
@@ -124,7 +135,7 @@
 separator: wpe_beamformer
 separator_conf:
     num_spk: 1
-    loss_type: spectrum
+    loss_type: spectrum_log
     use_wpe: false
     wnet_type: blstmp
     wlayers: 3
@@ -140,7 +151,7 @@
     bunits: 512
     bprojs: 512
     badim: 320
-    ref_channel: 4
+    ref_channel: 3
     use_noise_mask: true
     beamformer_type: mvdr_souden
     bdropout_rate: 0.0
@@ -150,5 +161,5 @@
     hop_length: 128
 required:
 - output_dir
-version: 0.10.3a3
+version: '202204'
 distributed: false
```

**Comment by Emrys365** [comment]
> I see. But I think the loss_type set in the separator (espnet2/enh/separator/nueral_beamformer.py) will not make any difference unless it is changed from 'spectrum' to 'mask_mse'. For the second one, I'll do an experiment to see if I can reproduce the original results. Thanks again!

For the second point, please also take care of the `ref_channel` argument in run.sh to make sure they have the same value.

**Comment by guoyifan97** [comment]
> > I see. But I think the loss_type set in the separator (espnet2/enh/separator/nueral_beamformer.py) will not make any difference unless it is changed from 'spectrum' to 'mask_mse'. For the second one, I'll do an experiment to see if I can reproduce the original results. Thanks again!
> 
> For the second point, please also take care of the `ref_channel` argument in run.sh to make sure they have the same value.

OK

**Comment by Emrys365** [comment]
@guoyifan97, I retrained a neural beamformer with the [default config](https://github.com/espnet/espnet/blob/master/egs2/chime4/enh1/conf/tuning/train_enh_beamformer_mvdr.yaml), and the results I obtained below looks comparable to previous ones, even without the loss on the noise mask.

|dataset|PESQ|STOI|SDR|SI_SNR|
|---|---|---|---|---|
|enhanced_dt05_simu_isolated_6ch_track|2.613|94.726|13.809|12.175|
|enhanced_et05_simu_isolated_6ch_track|2.647|95.192|15.707|14.241|

**Comment by guoyifan97** [comment]
@Emrys365 , Hi, thanks again. In fact, I have noticed that the model trained with the MSE loss on spectrum performs best when being decoded with ASR model (At least with CHiME4 data set and the ASR model trained on wsj corpus). And I am interested with the ASR performance.  So I pay more attention to the ASR metrics. 

And as mentioned last time, I have done another experiments with MSE loss on spec, and using the 4th mic as the reference channel just as the old espnet version. The results is as follows.

##############
new espnet version ('202204')
##############

| Reference Channel | Dataset | SDR | CER | WER |
| 3 | enhanced_dt05_simu_isolated_6ch_track | 15.51 | 14.7 | 29.0 |
| 3 | enhanced_et05_simu_isolated_6ch_track | 16.46 | 18.9 | 36.3 |
| 4 | enhanced_dt05_simu_isolated_6ch_track | 12.90 | 20.2 | 37.4 |
| 4 | enhanced_et05_simu_isolated_6ch_track | 12.99 | 26.5 | 47.5 |

###############
old version (0.10.3a3)
###############

| 4 | enhanced_dt05_simu_isolated_6ch_track | 14.77 |  | 37.4 |
| 4 | enhanced_et05_simu_isolated_6ch_track | 15.56| 26.5 | 47.5 |

**Comment by guoyifan97** [comment]
@Emrys365 , thanks again. I find the main reason for the issue is that I didn't change the reference channel option in the run.sh, while the configure file in the new version of the espnet changed the reference channel from 3 to 4 as you mentioned. After unifying the reference channel in run.sh and config file, the problem disappeared. Thanks again!
